### PR TITLE
Let qemu use /etc/resolv.conf of the target pod

### PR DESCRIPTION
**Pull request by rmohr**
QEMU already uses the network interfaces of the Pod. Also make the
/etc/resolv.conf of the container available for it.

This allows the QEMU process to resolve kubernetes DNS entries.

### CODE REVIEW

**Comment by rmohr** [comment]
/cc @fabiand @admiyo @stu-gott 

That principle works, will verify this script tomorrow.

**Comment by sungwonh** [comment]
@rmohr Could you tell me how QEMU uses the network interfaces of the Pod (virt-launcher)? I think QEMU uses the network interfaces of the Vagrant VM (master) in my development environment.

**Comment by rmohr** [comment]
@sungwonh You are right that qemu uses the libvirt network from the host, but it is in the network namespace of the pod. This works because libvirt is handing over all network connections to libvirt via file descriptors.

This means:
 - Libvirt networking can be used at the moment
 - Spice and VNC servers (managed by qemu directly) can listen to the pod network interfaces, which makes them discoverable through kubernetes

But, that also means that qemu still runs with the mount namespace of libvirt. Therefore it sees the `/etc/resolv.conf` which libvirt sees. If libvirt is not running inside a Pod itself, it sees the `/etc/resolv.conf` of the host. That basically means that kube dns is not working for qemu. In other words it can't look up services via kubernetes service discovery. One case where QEMU might want to do that, is connecting to iSCSI . This is what #94 by @fabiand wants to achieve.

**Comment by rmohr** [comment]
retest this please

**Comment by rmohr** [comment]
@stu-gott I think the example yamls and the test here are very valuable for you for #109

**Comment by rmohr** [comment]
retest this please

**Comment by rmohr** [comment]
retest this please

**Comment by rmohr** [comment]
retest this please

File: cluster/verify-qemu-kube [hunk 1]
```diff
@@ -56,23 +56,28 @@ for CG in $EXCLUDED_CGROUPS; do
...
-fi
+# Check namespaces
+echo ""
+echo "Checking for required namespaces:"
+NAMESPACES='pid net'
+for NS in $NAMESPACES; do
+  NS_PRETTY=`echo $NS | tr '[:lower:]' '[:upper:]'`
+  VM_PID_NS=$(ls -iL /proc/$VM_PID/ns/$NS | cut -f 1 -d " ")
+  CONTAINER_PID_NS=$(ls -iL /proc/$CONTAINER_PID/ns/$NS | cut -f 1 -d " ")
+
```
Comment (stu-gott) [inline cluster/verify-qemu-kube]: trailing whitespace

File: cluster/verify-qemu-kube [hunk 2]
```diff
@@ -56,23 +56,28 @@ for CG in $EXCLUDED_CGROUPS; do
...
+echo ""
+echo "Checking for required namespaces:"
+NAMESPACES='pid net'
+for NS in $NAMESPACES; do
+  NS_PRETTY=`echo $NS | tr '[:lower:]' '[:upper:]'`
+  VM_PID_NS=$(ls -iL /proc/$VM_PID/ns/$NS | cut -f 1 -d " ")
+  CONTAINER_PID_NS=$(ls -iL /proc/$CONTAINER_PID/ns/$NS | cut -f 1 -d " ")
+  
+  if [ "x$VM_PID_NS" == "x" ]; then
+    echo Could not determine $NS_PRETTY namespace of VM $VM_NAME
```
Comment (stu-gott) [inline cluster/verify-qemu-kube]: whitespace

File: cluster/verify-qemu-kube [hunk 3]
```diff
@@ -56,23 +56,28 @@ for CG in $EXCLUDED_CGROUPS; do
...
+for NS in $NAMESPACES; do
+  NS_PRETTY=`echo $NS | tr '[:lower:]' '[:upper:]'`
+  VM_PID_NS=$(ls -iL /proc/$VM_PID/ns/$NS | cut -f 1 -d " ")
+  CONTAINER_PID_NS=$(ls -iL /proc/$CONTAINER_PID/ns/$NS | cut -f 1 -d " ")
+  
+  if [ "x$VM_PID_NS" == "x" ]; then
+    echo Could not determine $NS_PRETTY namespace of VM $VM_NAME 
+    ERROR=1
+  fi
+
```
Comment (stu-gott) [inline cluster/verify-qemu-kube]: whitespace

File: cmd/virt-launcher/qemu-kube [hunk 4]
```diff
@@ -75,4 +75,9 @@ echo "$CMD" >> $LOG
 # Don't close file descriptors bigger than 10000 to allow passing tap device fds
 # Start the qemu process in the cgroups of the docker container
 # to adher the resource limitations of the container.
-sudo -C 10000 bash -c -- "cgclassify -g ${CGROUPS}:system.slice/docker-$CONTAINER_ID.scope --sticky \$\$ && nsenter -t $CONTAINER_PID -n -p $CMD"
+sudo -C 10000 unshare --mount bash -s << END
+  nsenter -m -t $CONTAINER_PID cat /etc/resolv.conf > /etc/resolv.conf.\$\$
```
Comment (fabiand) [inline cmd/virt-launcher/qemu-kube]: Just for reference - if libvirtd was running in a pod, then we wouldn't need to do this workaround.

Comment (rmohr) [inline cmd/virt-launcher/qemu-kube]: Correct :+1:

File: cluster/vm.json [hunk 5]
```diff
@@ -10,6 +10,28 @@
...
+            "source": {
+              "host": {
+                "name": "iscsi-demo-target",
+                "port": "3260"
+              },
+              "name": "iqn.2017-01.io.kubevirt:sn.42/2",
+              "protocol": "iscsi"
+            },
+            "target": {
+              "dev": "vda"
```
Comment (stu-gott) [inline cluster/vm.json]: Does this work by the way? On my system, qemu complains about "vda" but accepts "sda".

Comment (rmohr) [inline cluster/vm.json]: In the vagrant setup it works for me with vda/virtio,  hda/ide but not with sda/scsi.

File: cluster/verify-qemu-kube (hunk 1 above)
Comment (rmohr) [inline cluster/verify-qemu-kube]: done

File: cluster/verify-qemu-kube (hunk 2 above)
Comment (rmohr) [inline cluster/verify-qemu-kube]: done

File: cluster/verify-qemu-kube (hunk 3 above)
Comment (rmohr) [inline cluster/verify-qemu-kube]: done

**Comment by fabiand** [comment]
Merging with two acks.

**Comment by sungwonh** [comment]
@rmohr Could you tell me how to define "default" libvirt network which is refered to in vm.json to test this Pull Request?
Below is "default" libvirt network defined in the Vagrant VM (master). Is below definition right?

```bash
[root@master ~]# virsh net-dumpxml default
<network connections='1'>
  <name>default</name>
  <uuid>b5da8e9c-f433-41d8-b5c1-c5dcdb2c69a8</uuid>
  <forward mode='nat'>
    <nat>
      <port start='1024' end='65535'/>
    </nat>
  </forward>
  <bridge name='virbr0' stp='on' delay='0'/>
  <mac address='52:54:00:1a:dc:a1'/>
  <ip address='192.168.122.1' netmask='255.255.255.0'>
    <dhcp>
      <range start='192.168.122.2' end='192.168.122.254'/>
    </dhcp>
  </ip>
</network>
```

**Comment by rmohr** [comment]
@sungwonh this network looks fine. If your VM starts, it should have a network device which has an IP from this defined network range. You can define any other other libvirt network via `virsh` and connect your VMs to it too ...

Regarding to the resolv.conf, it will be used by qemu directly inside the pod, to connect to the lun described in the vm.json.

I am not entirely sure if this answers your question ...

**Comment by sungwonh** [comment]
@rmohr Thank you for your answer. Cloud you explain the following sentence which is from your previous answer in more detail?

> qemu uses the libvirt network from the host, but it is in the network namespace of the pod.

I think my qemu uses the libvirt network in the network namespace of the host (vagrant master) not the network namespace of the pod.

I also tested with the below VM spec, but my qemu still uses "eth0" in the network namespace of the host.

```
interfaces:
      - source:
          device: eth0
        type: direct
```

**Comment by rmohr** [comment]
> I also tested with the below VM spec, but my qemu still uses "eth0" in the network namespace of the host.

@sungwonh You are correct, that is the expected behaviour at the moment. I will try to rephrase.

So before we tell libvirt to start a VM, we have already started a Pod which is exclusively for that VM.
After this pod is running, we tell libvirt to start the VM via the `qemu-kube` script. This script takes all the qemu command-line parameters passed in by libvirt and starts the expected qemu process in the cgroups (all except devices cgroup) and namespaces (PID and NET) of the target pod. So the qemu process is now mostly running inside this pod.

As a consequence of this, qemu when it does it own network stuff (spice, vnc, raw attachment to luns, ...), interacts with the pod networks. However, if you are using libvirt to do networking (that includes the `direct` type), libvirt is actually opening a filedescriptor and passes that filedescriptor (from the tap, macvtap,...) to qemu.

In you concrete example:

1) Libvirt creates something like a macvtap, backed by eth0
2) Libvirt then opens a filedescriptor to the macvtap
3) Libvirt then passes this open filedescriptor to qemu

Therefore you can still use the libvirt networks for your VM interfaces, but qemu can be accessed by kubernetes infra through the pod nework and qemu can access kubernetes infra through the pod network.

**Comment by rmohr** [comment]
Oh and now with this patch, the qemu process also runs in its own mount namespace, which is based on libvirts mount namespace, but populated with the resolv.conf of the pod mount namespace, so that the qemu process has the right dns informations, matching the pod network namespace interfaces ...

**Comment by sungwonh** [comment]
@rmohr Thank you for your help. I was completely mistaken about this patch. I though this patch would enable access kubernetes infra using kubernetes dns in the guest VM.

**Comment by rmohr** [comment]
That would be nice if we would be that far already. We were discussing this topic with people from qemu and open shift at Devconf.cz this year. There are a few ideas floating around, like a special libvirt network backend, solving it by having libvirt in a container, integrate through ovn, ... but nothing too concrete yet.

@sungwonh Is that something you are especially interested in and where you would want to help?

**Comment by sungwonh** [comment]
@rmohr Yes, I am really interested in this topic. It would be great if I could contribute this topic.

**Comment by fabiand** [comment]
@sungwonh please chime into this thread https://groups.google.com/d/topic/kubevirt-dev/MluIvUZG2dI/discussion

**Pull request merged**